galaxy/management/commands/generate_galaxy_test_data.py

```diff
@@ -0,0 +1,288 @@
+        Page.objects.filter(name__startswith="galaxy-").delete()
+        User.objects.filter(username__startswith="galaxy-").delete()
+        Picture.objects.filter(name__startswith="galaxy-").delete()
+        Album.objects.filter(name__startswith="galaxy-").delete()
+        self.logger.info("Done. Populating a new Galaxy")
+
+        self.NB_USERS = options["user_pack_count"] * USER_PACK_SIZE
+        self.NB_CLUBS = options["club_count"]
+
+        self.now = timezone.now().replace(hour=12)
+        root = User.objects.filter(username="root").first()
+        sas = SithFile.objects.get(id=settings.SITH_SAS_ROOT_DIR_ID)
+        self.galaxy_album = Album.objects.create(
+            name="galaxy-register-file",
+            owner=root,
+            is_moderated=True,
+            is_in_sas=True,
+            parent=sas,
+        )
+
+        self.make_clubs()
+        self.make_users()
+        self.make_families()
+        self.make_club_memberships()
+        self.make_pictures()
+        self.make_pictures_memberships()
+        half_pack = USER_PACK_SIZE // 2
+        for u in range(half_pack, self.NB_USERS, half_pack):
+            self.make_important_citizen(u)
+
+    def make_clubs(self):
+        # This will create all the clubs and store them in self.clubs for fast access later
+        self.clubs = {}
+        for i in range(self.NB_CLUBS):
+            self.clubs[i] = Club.objects.create(
+                unix_name=f"galaxy-club-{i}", name=f"club-{i}"
+            )
+
+    def make_users(self):
+        # This will create all the users and store them in self.users for fast access later
+        self.users = {}
+        for i in range(self.NB_USERS):
+            u = User.objects.create_user(
+                username=f"galaxy-user-{i}",
+                email=f"{i}@galaxy.test",
+                first_name="Citizen",
+                last_name=f"n°{i}",
+            )
+            self.users[i] = u
+
+            self.logger.info(f"Registering {u}")
+            Subscription.objects.create(
+                member=u,
+                subscription_start=Subscription.compute_start(
+                    self.now - timedelta(days=self.NB_USERS - i)
+                ),
+                subscription_end=Subscription.compute_end(duration=2),
+            )
+
+    def make_families(self):
+        # This will iterate on all citizen after the 200th.
+        # Then it will take 14 other citizen among the 200 preceding (godfathers are usually older), and apply another heuristic to determine if they should have a family link
+        for i in range(200, self.NB_USERS):
+            for j in range(i - 200, i, 14):  # this will loop 14 times (14² = 196)
+                if (i / 10) % 10 == (i + j) % 10:
+                    u1 = self.users[i]
+                    u2 = self.users[j]
+                    self.logger.info(f"Making {u2} the godfather of {u1}")
+                    u1.godfathers.add(u2)
```

Comment:
This leads to a db hit for each family relationship instead of creating the godfathers all at once, which is not ideal performance-wise.

You should find all the godfathers of your user, and call `godfathers.add()` outside of the inner loop.
This could look like : 
```py
for i, user in enumerate(self.users[200:]): # assuming users are stored in a list
    godfathers = []
    for j in range(i - 200, i, 14):
        if (i / 10) % 10 == (i + j) % 10:
            godfathers.append(self.user[j])
            self.logger.info(f"Making {u2} the godfather of {u1}")
    user.godfathers.add(*godfathers)
```
You can also call `user.godfathers.set(godfathers)` if you want to make sure there is no family relationship but the ones you create in this function.

Reply:
Changed